plugins/src/gcc/imp.rs

```diff
@@ -0,0 +1,1274 @@
+            self.variance = (1. - MOVING_AVERAGE_SMOOTHING_FACTOR)
+                * (self.variance + MOVING_AVERAGE_SMOOTHING_FACTOR * avg_diff * avg_diff);
+            self.standard_dev = self.variance.sqrt();
+        }
+    }
+
+    fn estimate_is_close(&self, value: Bitrate) -> bool {
+        if self.average == 0. {
+            false
+        } else {
+            ((self.average - STANDARD_DEVIATION_CLOSE_NUM * self.standard_dev)
+                ..(self.average + STANDARD_DEVIATION_CLOSE_NUM * self.standard_dev))
+                .contains(&(value as f64))
+        }
+    }
+}
+
+struct State {
+    /// Note: The target bitrate applied is the min of
+    /// target_bitrate_on_delay and target_bitrate_on_loss
+    estimated_bitrate: Bitrate,
+
+    /// Bitrate target based on delay factor for all video streams.
+    /// Hasn't been tested with multiple video streams, but
+    /// current design is simply to divide bitrate equally.
+    target_bitrate_on_delay: Bitrate,
+
+    /// Used in additive mode to track last control time, influences
+    /// calculation of added value according to gcc section 5.5
+    last_increase_on_delay: Option<time::Instant>,
+    last_decrease_on_delay: time::Instant,
+
+    /// Bitrate target based on loss for all video streams.
+    target_bitrate_on_loss: Bitrate,
+
+    last_increase_on_loss: time::Instant,
+    last_decrease_on_loss: time::Instant,
+
+    /// Exponential moving average, updated when bitrate is
+    /// decreased
+    ema: ExponentialMovingAverage,
+
+    last_control_op: BandwidthEstimationOp,
+
+    min_bitrate: Bitrate,
+    max_bitrate: Bitrate,
+
+    detector: Detector,
+
+    periodic_push: Option<gst::PeriodicClockId>,
+
+    // Implemented like a leaky bucket
+    buffers: VecDeque<gst::Buffer>,
+
+    flow_return: Result<gst::FlowSuccess, gst::FlowError>,
+    last_push: time::Instant,
+}
+
+impl Default for State {
+    fn default() -> Self {
+        Self {
+            target_bitrate_on_delay: DEFAULT_ESTIMATED_BITRATE,
+            target_bitrate_on_loss: DEFAULT_ESTIMATED_BITRATE,
+            last_increase_on_loss: time::Instant::now(),
+            last_decrease_on_loss: time::Instant::now(),
+            ema: Default::default(),
+            last_increase_on_delay: None,
+            last_decrease_on_delay: time::Instant::now(),
+            min_bitrate: DEFAULT_MIN_BITRATE,
+            max_bitrate: DEFAULT_MAX_BITRATE,
+            detector: Detector::new(),
+            buffers: Default::default(),
+            estimated_bitrate: DEFAULT_ESTIMATED_BITRATE,
+            last_control_op: BandwidthEstimationOp::Increase("Initial increase".into()),
+            flow_return: Err(gst::FlowError::Flushing),
+            periodic_push: None,
+            last_push: time::Instant::now(),
+        }
+    }
+}
+
+impl State {
+    fn create_buffer_list(&mut self, bwe: &super::BandwidthEstimator) -> gst::BufferList {
```

Comment:
Add a comment here pointing out that this is implementing `4. sending engine`, using a "leaky bucket" implementation?